mlflow/pyfunc/scoring_server/client.py

```diff
@@ -0,0 +1,71 @@
+import requests
+import time
+from mlflow.pyfunc import scoring_server
+import subprocess
+import os
+import sys
+import signal
+import json
+import atexit
+import warnings
+
+
+class ScoringServerClient:
+    def __init__(self, host, port):
+        self.url_prefix = f"http://{host}:{port}"
+
+    def ping(self):
+        ping_status = requests.get(url=self.url_prefix + "/ping")
+        if ping_status.status_code != 200:
+            raise Exception(f"ping failed (error code {ping_status.status_code})")
+
+    def wait_server_ready(self, timeout=30):
+        begin_time = time.time()
+        while True:
+            time.sleep(0.3)
+            try:
+                self.ping()
+                return
+            except Exception:
+                pass
+            if time.time() - begin_time > timeout:
+                break
+        raise RuntimeError("Wait scoring server ready timeout.")
+
+    def invoke(self, data):
+        """
+        Invoke inference on input data. The input data must be pandas dataframe or json instance.
+        """
+
+        import pandas as pd
+
+        if isinstance(data, pd.DataFrame):
+            content_type = scoring_server.CONTENT_TYPE_JSON_RECORDS_ORIENTED
+        else:
+            content_type = scoring_server.CONTENT_TYPE_JSON
+
+        post_data = json.dumps(scoring_server._get_jsonable_obj(data, pandas_orient="records"))
+
+        response = requests.post(
+            url=self.url_prefix + "/invocations",
+            data=post_data,
+            headers={"Content-Type": content_type},
+        )
+
+        if response.status_code != 200:
+            raise Exception(
+                f"Invocation failed (error code {response.status_code}, response: {response.text})"
+            )
+
+        return scoring_server.infer_and_parse_json_input(response.text)
+
+
+def prepare_env(local_model_path, stdout=sys.stdout, stderr=sys.stderr):
+    cmd = [
+        "mlflow",
+        "models",
+        "prepare-env",
+        "-m",
+        local_model_path,
+    ]
+    return subprocess.run(cmd, stdout=stdout, stderr=stderr, universal_newlines=True, check=True)
```

Comment:
This failed when I tried it on Databricks with the following stacktrace:

```
UnsupportedOperation                      Traceback (most recent call last)
<[command-478391771928007]()> in <module>
----> 1 udf = mlflow.pyfunc.spark_udf(spark, logged_model.model_uri, env_manager="conda")

/local_disk0/.ephemeral_nfs/envs/pythonEnv-bec7f63d-3c4f-4a99-94e4-ee6b73674b94/lib/python3.9/site-packages/mlflow/pyfunc/__init__.py in spark_udf(spark, model_uri, result_type, env_manager)
    944         # Prepare restored environment in driver side if possible.
    945         if env_manager == "conda":
--> 946             prepare_env(local_model_path)
    947 
    948     model_metadata = Model.load(os.path.join(local_model_path, MLMODEL_FILE_NAME))

/local_disk0/.ephemeral_nfs/envs/pythonEnv-bec7f63d-3c4f-4a99-94e4-ee6b73674b94/lib/python3.9/site-packages/mlflow/pyfunc/scoring_server/client.py in prepare_env(local_model_path, stdout, stderr)
     69         local_model_path,
     70     ]
---> 71     return subprocess.run(cmd, stdout=stdout, stderr=stderr, universal_newlines=True, check=True)

/usr/lib/python3.9/subprocess.py in run(input, capture_output, timeout, check, *popenargs, **kwargs)
    503         kwargs['stderr'] = PIPE
    504 
--> 505     with Popen(*popenargs, **kwargs) as process:
    506         try:
    507             stdout, stderr = process.communicate(input, timeout=timeout)

/usr/lib/python3.9/subprocess.py in __init__(self, args, bufsize, executable, stdin, stdout, stderr, preexec_fn, close_fds, shell, cwd, env, universal_newlines, startupinfo, creationflags, restore_signals, start_new_session, pass_fds, user, group, extra_groups, encoding, errors, text, umask)
    827         (p2cread, p2cwrite,
    828          c2pread, c2pwrite,
--> 829          errread, errwrite) = self._get_handles(stdin, stdout, stderr)
    830 
    831         # We wrap OS handles *before* launching the child, otherwise a

/usr/lib/python3.9/subprocess.py in _get_handles(self, stdin, stdout, stderr)
   1596             else:
   1597                 # Assuming file-like object
-> 1598                 c2pwrite = stdout.fileno()
   1599 
   1600             if stderr is None:

/databricks/python/lib/python3.9/site-packages/ipykernel/iostream.py in fileno(self)
    306             return self._original_stdstream_copy
    307         else:
--> 308             raise io.UnsupportedOperation("fileno")
    309 
    310     def _watch_pipe_fd(self):

UnsupportedOperation: fileno
```

Do we need to run this in a subprocess with `check=True`, or can we call `PyFuncBackend.prepare_env()` directly?

Reply:
Replaced with:
```
            _get_flavor_backend(local_model_path, no_conda=False, install_mlflow=False).prepare_env(
                model_uri=local_model_path
            )
```